### PR TITLE
fix(argo-cd): make automountServiceAccountToken configurable

### PR DESCRIPTION
**Pull request by stefan-caraiman**
Keeps the default behaviour of having automountServiceAccountToken set to true for deployments/statefulsets.

Checklist:

* [x] I have bumped the chart version according to [versioning](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#versioning)
* [x] I have updated the documentation according to [documentation](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#documentation)
* [x] I have updated the chart changelog with all the changes that come with this pull request according to [changelog](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#changelog).
* [x] Any new values are backwards compatible and/or have sensible default.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md).
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/stable/developer-guide/ci/)).

### CODE REVIEW

**Comment by mkilchhofer** [comment]
Just to probably learn something new: is there a difference setting it on the deployment/statefulset or ServiceAccount?
Because this toggle is already supported on all ServiceAccounts:

```markdown
$ grep automount README.md
| controller.serviceAccount.automountServiceAccountToken | bool | `true` | Automount API credentials for the Service Account |
| repoServer.serviceAccount.automountServiceAccountToken | bool | `true` | Automount API credentials for the Service Account |
| server.serviceAccount.automountServiceAccountToken | bool | `true` | Automount API credentials for the Service Account |
| dex.serviceAccount.automountServiceAccountToken | bool | `true` | Automount API credentials for the Service Account |
| redis.serviceAccount.automountServiceAccountToken | bool | `false` | Automount API credentials for the Service Account |
| applicationSet.serviceAccount.automountServiceAccountToken | bool | `true` | Automount API credentials for the Service Account |
| notifications.serviceAccount.automountServiceAccountToken | bool | `true` | Automount API credentials for the Service Account |
```

**Comment by stefan-caraiman** [comment]
> Just to probably learn something new: is there a difference setting it on the deployment/statefulset or ServiceAccount? Because this toggle is already supported on all ServiceAccounts:
> 
> ```
> $ grep automount README.md
> | controller.serviceAccount.automountServiceAccountToken | bool | `true` | Automount API credentials for the Service Account |
> | repoServer.serviceAccount.automountServiceAccountToken | bool | `true` | Automount API credentials for the Service Account |
> | server.serviceAccount.automountServiceAccountToken | bool | `true` | Automount API credentials for the Service Account |
> | dex.serviceAccount.automountServiceAccountToken | bool | `true` | Automount API credentials for the Service Account |
> | redis.serviceAccount.automountServiceAccountToken | bool | `false` | Automount API credentials for the Service Account |
> | applicationSet.serviceAccount.automountServiceAccountToken | bool | `true` | Automount API credentials for the Service Account |
> | notifications.serviceAccount.automountServiceAccountToken | bool | `true` | Automount API credentials for the Service Account |
> ```

Yep 😄  as per the [K8s docs](https://v1-24.docs.kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#use-the-default-service-account-to-access-the-api-server):

```
The pod spec takes precedence over the service account if both specify a automountServiceAccountToken value.
```

Setting automountServiceAccountToken on a ServiceAccount applies the setting to all pods using that account, defaulting them to either mount or not mount the service account token. Setting it on a Deployment or StatefulSet allows for specific control over that particular workload, overriding the ServiceAccount's default.

**Comment by mkilchhofer** [comment]
> Setting automountServiceAccountToken on a ServiceAccount applies the setting to all pods using that account, defaulting them to either mount or not mount the service account token. Setting it on a Deployment or StatefulSet allows for specific control over that particular workload, overriding the ServiceAccount's default.

Ok but do you create your own SA's or why do you like this change? By default every deployment/component of Argo CD has its own SA :)

**Comment by stefan-caraiman** [comment]
Not really no, we do not create our own SA, we rely on the defaults, but due to the following Azure AKS Security policy:
```
"Kubernetes clusters should disable automounting API credentials" - Disable automounting API credentials to prevent a potentially compromised Pod resource to run API commands against Kubernetes clusters.
```

TLDR: We need to basically set `automountServiceAccountToken: false` to both the SA and Deployments/StatefulSets for ArgoCD and then mount the `serviceaccounttoken` through extra `volumes/volumeMounts` for resources that do need access to the API.

Example [values](https://cert-manager.io/docs/installation/best-practice/#best-practice-helm-chart-values) & [issue](https://github.com/cert-manager/cert-manager/issues/5254) from cert-manager on this topic.

In the end the result is the same as with `automountServiceAccountToken: true`, but it is a "hardening" requirement to be compliant with the Azure Security Score 😅

**Comment by mkilchhofer** [comment]
> In the end the result is the same as with `automountServiceAccountToken: true`, but it is a "hardening" requirement to be compliant with the Azure Security Score 😅

Oh boy 🙈 . I now definitively learned something new. But IMHO this is only to satisfy the security scanner, it doesn't really improve security.
There was a good talk at KubeCon: Malicious compliance ;)
[IMAGE ALT TEXT HERE](https://www.youtube.com/watch?v=9weGi0csBZM)

Lets wait on some other opinions of my colleagues.

**Comment by stefan-caraiman** [comment]
Yep totally agree. No harm though in adding in the option to customize it since it defaults to true just as before. 😄

**Comment by mbevc1** [comment]
Hah, nice work Microsoft :)

Good reference @mkilchhofer :+1: 

Happy we add it in for people on Azure as long as the defaults are not changing.

**Comment by stefan-caraiman** [comment]
@mbevc1 indeed, no defaults are changing so the change is harmless 👍

**Comment by stefan-caraiman** [comment]
@mbevc1 @mkilchhofer just to exemplify further, the following 2 example `values` achieve the same injection of the Service account token into the pod in order to make K8s api calls.

```
# Example 1 ( the default now and with this MR too)
# Automatically injection done as explained here
# https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#serviceaccount-admission-controller
controller:
  automountServiceAccountToken: true
  serviceAccount:
    automountServiceAccountToken: true
```
same as with
```
# Example 2 values for disabling auto-mounting and opting to inject it via volumes instead
# for "hardened" AKS setups
controller:
  automountServiceAccountToken: false
  serviceAccount:
    automountServiceAccountToken: false
  volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: serviceaccount-token
      readOnly: true
  volumes:
    - name: serviceaccount-token
      projected:
        defaultMode: 0444
        sources:
        - serviceAccountToken:
            expirationSeconds: 3607
            path: token
        - configMap:
            name: kube-root-ca.crt
            items:
            - key: ca.crt
              path: ca.crt
        - downwardAPI:
            items:
            - path: namespace
              fieldRef:
                apiVersion: v1
                fieldPath: metadata.namespace
```

**Review by yu-croco — APPROVED** [review]
Thank you for your contribution. LGTM.

**Review by mbevc1 — APPROVED** [review]

**Comment by mbevc1** [comment]
Thanks @stefan-caraiman

**Review by mkilchhofer — APPROVED** [review]
